server/src/main/java/org/elasticsearch/search/query/QueryPhase.java

```diff
@@ -59,6 +64,71 @@ public class QueryPhase {
     public QueryPhase() {}
 
     public static void execute(SearchContext searchContext) throws QueryPhaseExecutionException {
+        if (searchContext.rankShardContext() == null) {
+            executeQuery(searchContext);
+        } else {
+            executeRank(searchContext);
+        }
+    }
+
+    static void executeRank(SearchContext searchContext) throws QueryPhaseExecutionException {
+        try {
+            RankShardContext rankShardContext = searchContext.rankShardContext();
+            QuerySearchResult querySearchResult = searchContext.queryResult();
+            RankSearchContext rankSearchContext = new RankSearchContext(searchContext);
+
+            // run the combined boolean query total hits or aggregations
+            // otherwise mark top docs as empty
+            if (searchContext.suggest() != null
+                || searchContext.trackTotalHitsUpTo() != TRACK_TOTAL_HITS_DISABLED
+                || searchContext.aggregations() != null) {
+                QueryPhase.executeQuery(rankSearchContext);
+            } else {
+                searchContext.queryResult()
+                    .topDocs(
+                        new TopDocsAndMaxScore(
+                            new TopDocs(new TotalHits(0, TotalHits.Relation.EQUAL_TO), Lucene.EMPTY_SCORE_DOCS),
+                            Float.NaN
+                        ),
+                        new DocValueFormat[0]
+                    );
+            }
+
+            List<TopDocs> rrfRankResults = new ArrayList<>();
+            rankSearchContext.windowSize(rankShardContext.windowSize());
+            boolean searchTimedOut = querySearchResult.searchTimedOut();
+            long serviceTimeEWMA = querySearchResult.serviceTimeEWMA();
+            int nodeQueueSize = querySearchResult.nodeQueueSize();
+
+            // run each of the rrf queries
+            for (Query query : rankShardContext.queries()) {
+                // if a search timeout occurs, exit with partial results
+                if (searchTimedOut) {
+                    break;
+                }
+                rankSearchContext.rankQuery(query);
+                QueryPhase.executeQuery(rankSearchContext);
```

Comment:
it seems to me that we don't support any feature within executeQuery other than what executeWithCollectors does, hence we could call such method directly for clarity?

Reply:
I think that's true for now. Obviously, this may change as we add more feature support for RRF, but I'll go ahead and make this change.

Reply:
Agreed, we will have to change it but then it's going to be more visible that we need to do so.